monitoring/influxdb.go

```diff
@@ -105,52 +210,141 @@ type influxdbMonitor struct {
-	if fields == nil {
-		fields = map[string]interface{}{}
+type cacheEvent struct {
+	cacheEvents        int
+	maxCacheEvents     int
+	currentCacheEvents int
+}
+
+func newCacheEvent(cacheEvents int, maxCacheEvents int) *cacheEvent {
+	return &cacheEvent{
+		cacheEvents:        cacheEvents,
+		maxCacheEvents:     maxCacheEvents,
+		currentCacheEvents: cacheEvents,
 	}
+}
 
-	fields["value"] = value
+func (ce *cacheEvent) CurrentCacheEvents() int {
+	return ce.currentCacheEvents
+}
+
+func (ce *cacheEvent) IncreaseCacheEvents() {
+	currEvents := ce.currentCacheEvents + ce.cacheEvents
+	if currEvents > ce.maxCacheEvents {
+		ce.currentCacheEvents = ce.maxCacheEvents
+	} else {
+		ce.currentCacheEvents = currEvents
+	}
+}
+
+func (ce *cacheEvent) Reset() {
+	// This if is rarely true, so add it.
+	if ce.currentCacheEvents != ce.cacheEvents {
+		ce.currentCacheEvents = ce.cacheEvents
+	}
+}
+
+func (im influxdbMonitor) batchWriteTicker() {
+	var points []*influxdb.Point
+	t := time.NewTicker(im.batchWriteInterval)
+
+	for {
+		select {
+		case <-t.C:
+			im.batchWriteAndCheckErr(&points)
+
+		case pt := <-im.pointChan:
+			points = append(points, pt)
+
+			if len(points) >= im.cacheEvent.CurrentCacheEvents() {
+				im.batchWriteAndCheckErr(&points)
+			}
+		}
+	}
+}
+
+func (im influxdbMonitor) batchWriteAndCheckErr(points *[]*influxdb.Point) {
+	err := im.batchWrite(points)
+	if err != nil {
+		im.cacheEvent.IncreaseCacheEvents()
+
+		if len(*points) >= im.cacheEvent.CurrentCacheEvents() {
+			*points = nil
+			_ = im.logger.Error().Log(
+				"msg", "influxdb write failed and event number reach max-cache-events, cache events was cleaned up",
+			)
+		}
+	} else {
+		im.cacheEvent.Reset()
+	}
+}
+
+// *points will be set to nil if write successful.
+func (im influxdbMonitor) batchWrite(points *[]*influxdb.Point) error {
+	if points == nil || len(*points) == 0 {
+		return nil
+	}
 
 	bp, err := influxdb.NewBatchPoints(influxdb.BatchPointsConfig{
 		Database: im.database,
 	})
-
-	l := im.logger.With("database", im.database,
-		"measurement", measurement,
-		"value", value,
-		"tags", tags)
-
 	if err != nil {
-		l.Error().Log(
+		_ = im.logger.Error().Log(
+			"database", im.database,
 			"err", err,
 			"during", "influxdb.NewBatchPoints",
-			"msg", fmt.Sprintf("Error initializing batch points for %s: %v", measurement, err),
+			"msg", fmt.Sprintf("NewBatchPoints failed: %v", err),
 		)
+		return nil
 	}
 
-	pt, err := influxdb.NewPoint(measurement, tags, fields, at)
+	bp.AddPoints(*points)
 
+	err = im.client.Write(bp)
 	if err != nil {
-		l.Error().Log(
+		_ = im.logger.Error().Log(
+			"database", im.database,
 			"err", err,
-			"during", "influxdb.NewPoint",
-			"msg", fmt.Sprintf("Error initializing a point for %s: %v", measurement, err),
+			"during", "influxdb.client.Write",
+			"msg", fmt.Sprintf("influxdb client write points failed: %v", err),
 		)
+		return errors.Errorf("influxdb client write points failed: %v", err)
```

Comment:
Please use `errors.Wrap`.

Reply:
Fixed in https://github.com/theplant/appkit/blob/70c4f66134115cb94f1b703e0807a09a305c5fb5/monitoring/influxdb.go#L315.